onnxruntime/core/providers/webgpu/math/cum_sum.cc

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/webgpu/math/cum_sum.h"
+#include "core/providers/webgpu/shader_helper.h"
+#include "core/providers/webgpu/webgpu_supported_types.h"
+
+namespace onnxruntime {
+namespace webgpu {
+
+ONNX_OPERATOR_VERSIONED_KERNEL_EX(
+    CumSum,
+    kOnnxDomain,
+    11, 13,
+    kWebGpuExecutionProvider,
+    (*KernelDefBuilder::Create()).TypeConstraint("T", WebGpuSupportedFloatTypes()).TypeConstraint("T2", WebGpuSupportedNumberTypes()),
+    CumSum);
+
+ONNX_OPERATOR_KERNEL_EX(
+    CumSum,
+    kOnnxDomain,
+    14,
+    kWebGpuExecutionProvider,
+    (*KernelDefBuilder::Create()).TypeConstraint("T", WebGpuSupportedFloatTypes()).TypeConstraint("T2", WebGpuSupportedNumberTypes()),
+    CumSum);
+
+Status CumSumProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  const ShaderVariableHelper& input = shader.AddInput("input", ShaderUsage::UseUniform);
+  const ShaderVariableHelper& output = shader.AddOutput("output", ShaderUsage::UseUniform);
+
+  std::string index = "i32(" + input.IndicesGet("input_indices", "uniforms.axis") + ")";
+  std::string max = GetElementAt("uniforms.input_shape", "uniforms.axis", input.Rank());
+  std::string lowerLimit = reverse_ ? index + (exclusive_ ? " + 1" : "") : "0";
+  std::string upperLimit = reverse_ ? max : index + (exclusive_ ? "" : " + 1");
+  shader.MainFunctionBody() << shader.GuardAgainstOutOfBoundsWorkgroupSizes("uniforms.output_size")
+                            << "var input_indices = " << input.OffsetToIndices("global_idx") << ";\n"
+                            << "var sum = output_indices_t(0);\n"
+                            << "let first : i32 = " << lowerLimit << ";\n"
+                            << "let last : i32 = " << upperLimit << ";\n"
+                            << "for (var i : i32 = first; i < last; i++) {\n"
+                            << "  " << input.IndicesSet("input_indices", "uniforms.axis", "u32(i)") << ";\n"
+                            << "  sum = sum + " << input.GetByIndices("input_indices") << ";\n"
+                            << "}\n"
+                            << output.SetByOffset("global_idx", "sum") << ";\n";
+
+  return Status::OK();
+}
+
+Status CumSum::ComputeInternal(ComputeContext& context) const {
+  const auto* input_tensor = context.Input(0);
+  const TensorShape& input_shape = input_tensor->Shape();
+  int64_t input_rank = input_shape.NumDimensions();
+
+  const auto* axis_tensor = context.Input(1);
+  const auto* axis_data = axis_tensor->Data<int64_t>();
+
+  // Handle negative axis
+  int64_t axis = axis_data[0];
```

Comment:
Need to check if axis is valid, i.e. in the valid range `-input_range <= axis` and `axis < input_range`.